### PR TITLE
request: implement multiple request pool

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Extend handle object for request objects to have multiple pools.
Multiple pools allows request objects to be allocated and freed in
parallel, critical for threading performance.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)\
      PR #4273
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
@sagarth I don't quite like PR #4273 where it intrusively added extra pool bits to all handle objects even when most of them would never need to use the pool. So I created this PR for a "lighter" weight implementation. Could you take a look?

File: src/mpid/ch4/netmod/ofi/ofi_types.h [hunk 1]
```diff
@@ -502,7 +496,6 @@ typedef struct {
 
 typedef struct MPIDI_OFI_win_request {
     MPIR_OBJECT_HEADER;
-    char pad[MPIDI_REQUEST_HDR_SIZE - MPIDI_OFI_OBJECT_HEADER_SIZE];
     struct fi_context context[MPIDI_OFI_CONTEXT_STRUCTS];       /* fixed field, do not move */
```
Comment (sagarth) [inline src/mpid/ch4/netmod/ofi/ofi_types.h]: @hzhou What was the reason that you removed the padding?

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: It appears that `MPIDI_OFI_win_request` is simply an OFI internal object that gets passed around and it is not a regular `MPIR_Request` at all. I have refactored for it to have its own object allocation using `MPIR_INTERNAL` kind type; so I believe the padding is no longer needed.

**Comment by hzhou** [comment]
test: jenkins/ch3/most
test: jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpid/ch4/netmod/ofi/ofi_types.h (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h]: In fact, we could simply use `MPL_malloc` and `MPL_free` for `MPIDI_OFI_win_request` objects :)

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

File: src/include/mpir_request.h [hunk 2]
```diff
@@ -357,9 +357,6 @@ static inline MPIR_Request *MPIR_Request_create(MPIR_Request_kind_t kind, int po
 #define MPIR_Request_add_ref(req_p_) \
     do { MPIR_Object_add_ref(req_p_); } while (0)
 
-#define MPIR_Request_release_ref(req_p_, inuse_) \
```
Comment (raffenet) [inline src/include/mpir_request.h]: We have a lot of these wrappers, though not consistently applied to all object types. For instance group, attr, comm, errhandler, and datatype all have some version of add/release_ref macros. I think a cleanup is worthwhile, but it should be more complete. Perhaps to defer this in order to stay on fast track.

File: src/mpi/coll/transports/gentran/tsp_gentran.c [hunk 3]
```diff
@@ -409,7 +409,7 @@ int MPII_Genutil_sched_start(MPII_Genutil_sched_t * sched, MPIR_Comm * comm, MPI
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPII_GENUTIL_SCHED_START);
 
     /* Create a request */
-    reqp = MPIR_Request_create(MPIR_REQUEST_KIND__COLL);
+    reqp = MPIR_Request_create(MPIR_REQUEST_KIND__COLL, 0);
```
Comment (raffenet) [inline src/mpi/coll/transports/gentran/tsp_gentran.c]: Will a pool be specified eventually by all components creating requests? For instance, in this example will collective requests become pool aware? I'm trying to wrap my head around the request creation API change.

What I'm getting at is that if request pools will only be used by ch4, then I wonder if `MPIR_Request_create` should stay as it is by hardcoding pool 0, and a separate pool-aware create/free API be added. Of course developers will need to understand when to use the new API. Maybe that's too complex. I'm open to what others think...

File: src/include/mpir_objects.h [hunk 4]
```diff
@@ -274,18 +274,17 @@ typedef int Handle_ref_count;
...
 #elif MPICH_THREAD_REFCOUNT == MPICH_REFCOUNT__LOCKFREE
 
-#include "opa_primitives.h"
-typedef OPA_int_t Handle_ref_count;
+typedef MPL_atomic_int_t Handle_ref_count;
 
 #define MPIR_Object_set_ref(objptr_,val)                        \
     do {                                                        \
-        OPA_store_int(&(objptr_)->ref_count, val);              \
+        MPL_atomic_store_int(&(objptr_)->ref_count, val); \
```
Comment (raffenet) [inline src/include/mpir_objects.h]: Let's split this to another PR and just focus on the pool changes.

File: src/mpid/ch4/netmod/ofi/ofi_rma.h [hunk 5]
```diff
@@ -217,7 +217,7 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_allocate_win_request_put_get(MPIR_Win * w
         + MPIDI_OFI_align_iov_len(alloc_iovs * t_size)
         + MPIDI_OFI_IOVEC_ALIGN - 1;    /* in case iov_store[0] is not aligned as we want */
 
-    req = (MPIDI_OFI_win_request_t *) MPIR_Request_create(MPIR_REQUEST_KIND__RMA, 0);
+    req = MPIDI_OFI_win_request_create();
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ofi/ofi_rma.h]: I *really* like the win_request changes, but it's another one I think is not critical to the request pool implementation, right? I'd prefer again to split it out and review it separately (along with someone from Intel).

File: src/include/mpir_request.h [hunk 6]
```diff
@@ -204,11 +204,59 @@ struct MPIR_Request {
...
+#define REQUEST_OBJECT_MAX   4096
+
+#define MPIR_REQUEST_NUM_POOLS REQUEST_POOL_MAX
 #define MPIR_REQUEST_PREALLOC 8
 
-extern MPIR_Object_alloc_t MPIR_Request_mem;
-/* Preallocated request objects */
-extern MPIR_Request MPIR_Request_direct[];
+extern MPIR_Object_alloc_t MPIR_Request_mem[MPIR_REQUEST_NUM_POOLS];
+extern MPIR_Request MPIR_Request_direct[MPIR_REQUEST_NUM_POOLS][MPIR_REQUEST_PREALLOC];
```
Comment (raffenet) [inline src/include/mpir_request.h]: This will have some impact on memory footprint (8 -> 512 direct requests). We might need to be more conservative. With ch3, `MPIR_Request` size is ~1k. ch4:ofi is 328 bytes.

File: src/include/mpir_objects.h (hunk 4 above)
Comment (hzhou) [reply src/include/mpir_objects.h]: OK.

File: src/include/mpir_request.h (hunk 6 above)
Comment (hzhou) [reply src/include/mpir_request.h]: I can add some code to make direct-pool only available for vci 0 -- to reduce the impact.

File: src/include/mpir_request.h (hunk 2 above)
Comment (hzhou) [reply src/include/mpir_request.h]: OK

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 3 above)
Comment (hzhou) [reply src/mpi/coll/transports/gentran/tsp_gentran.c]: Pool need be specified by all components. However, component can always choose pool 0 (or any specific pool) and will always be correct -- just mean they will be contending on that pool. Eventually, I think the collectives will also benefit from per-vci locking. For example, user can designate certain collective on specific vci to avoid contention with other threads' pt2pt communications.

Your suggestions could work. However, all requests creation is in a fast path and having explicit pool (vci) argument is good for raising the awareness (of developer). Especially now I created separated paths for request objects from rest of the handlemem allocation, and they may further deviate in the future, so I think let request have distinct interface is not a bad idea.

File: src/include/mpir_request.h (hunk 6 above)
Comment (raffenet) [reply src/include/mpir_request.h]: This sounds reasonable to do for now.

File: src/mpi/coll/transports/gentran/tsp_gentran.c (hunk 3 above)
Comment (raffenet) [reply src/mpi/coll/transports/gentran/tsp_gentran.c]: OK, it's good to know that eventually it would make sense for all components to be pool-aware. If others agree, then I think the API change should be OK to merge.

**Comment by hzhou** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by hzhou** [comment]
test:jenkins/ch4/ofi
